src/constraint_system/zelbet.rs

```diff
@@ -0,0 +1,1017 @@
+        self.range_gate(difference, 10);
+
+        // x' = x_1 * s_2 + x_2, this is the start of the composition
+        let mut acc = self.big_mul(
+            BlsScalar::one(),
+            nibbles[26],
+            s_ik_var,
+            Some((BlsScalar::one(), nibbles[25])),
+            BlsScalar::zero(),
+            BlsScalar::zero(),
+        );
+
+        (1..26).for_each(|k| {
+            let s_ik_var = self.add_input(BlsScalar::from_raw(s_i[25 - k].0));
+            let difference = self.big_add(
+                (BlsScalar::one(), s_ik_var),
+                (-BlsScalar::one(), nibbles_montgomery[25 - k]),
+                Some((-BlsScalar::one(), one)),
+                BlsScalar::zero(),
+                BlsScalar::zero(),
+            );
+            self.range_gate(difference, 12);
+            acc = self.big_mul(
+                BlsScalar::one(),
+                acc,
+                s_ik_var,
+                Some((BlsScalar::one(), nibbles[25 - k])),
+                BlsScalar::zero(),
+                BlsScalar::zero(),
+            );
+        });
+
+        self.constrain_to_constant(acc, reduced_input, BlsScalar::zero());
+
+        nibbles
+    }
+
+    /// S-box using hash tables
+    /// Assumes input BlsScalar value is in reduced form,
+    /// but outputs the result in Montgomery
+    pub fn s_box(&mut self, input: Variable) -> Variable {
+        let value = u256(self.variables[&input].0);
+        let permutation = match value < u256([659, 0, 0, 0]) {
+            true => BlsScalar(SBOX_BLS[value.0[0] as usize].0),
+            false => BlsScalar(value.0),
+        };
+
+        // let permutation_var =
+        // self.add_input(BlsScalar::from_raw(permutation.0));
+        // self.plookup_gate(input, input, permutation_var, None,
+        // BlsScalar::zero())
+        self.add_input(BlsScalar::from_raw(permutation.0))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::super::helper::*;
+    use super::*;
+    use crate::{constraint_system::StandardComposer, plookup::PlookupTable3Arity};
+    use dusk_bls12_381::BlsScalar;
+
+    #[test]
+    fn test_decomposition() {
+        let res = gadget_tester(
+            |composer| {
+                let one = composer.add_input(BlsScalar::one());
+                let output = composer.decomposition_gadget(one, DECOMPOSITION_S_I, INVERSES_S_I);
+                (1..27).for_each(|k| {
+                    composer.constrain_to_constant(output[k], BlsScalar::zero(), BlsScalar::zero());
+                });
+                // Check x_27 = 1, bearing in mind that x_1 is not in Montgomery form
+                composer.constrain_to_constant(
+                    output[0],
+                    BlsScalar::one().reduce(),
+                    BlsScalar::zero(),
+                );
+            },
+            800,
+        );
+        assert!(res.is_ok());
+    }
+}
+
+/// decomposition = [s_n, s_{n-1} ..., s_1]
+pub const DECOMPOSITION_S_I: [BlsScalar; 27] = [
```

Comment:
Move these constants into their own file, just a bit easier to get around that way :)